### PR TITLE
Hook system debug messages

### PR DESCRIPTION
**Pull request by mikee47**
To help track down "E:M" errors output by system functions, #define UART_DEBUG_SNOOP and rebuild. Any lines starting with "E:M" will then receive a full stack dump.

### CODE REVIEW

**Comment by mikee47** [comment]
Another way to do this is to put the code into your application, then call `ets_install_putc1` on that. Saves having to rebuild the framework. Thoughts?

File: Sming/system/uart.cpp [hunk 1]
```diff
@@ -718,11 +718,61 @@ void uart_set_pins(uart_t* uart, int tx, int rx)
...
+	if (uart) {
 		uart_write_char(uart, c);
+
+#ifdef UART_DEBUG_SNOOP
+		static char line[16];
+		static unsigned lineLength = 0;
+
+		if(c == '\r' || c == '\n') {
+			if(lineLength >= 2) {
+				if(memcmp(line, "E:M", 2) == 0) {
```
Comment (slaff) [inline Sming/system/uart.cpp]: @mikee47 Maybe you can change this to "E:M " ( <-- one space after the capital M) just to avoid false positives.

**Comment by slaff** [comment]
> Another way to do this is to put the code into your application, then call `ets_install_putc1` on that. Saves having to rebuild the framework. Could also trap text other than "E:M". Thoughts?

Can you paste here some code to illustrate also the other approach?

**Comment by mikee47** [comment]
> > Another way to do this is to put the code into your application, then call `ets_install_putc1` on that. Saves having to rebuild the framework. Could also trap text other than "E:M". Thoughts?
> Can you paste here some code to illustrate also the other approach?

OK, so instead of changing `uart.cpp` we could place a version of `uart_debug_putc()` into a separate module, maybe just add it to `crash_handler.c` where we already have an implementation of `print_stack()`. Something like:


```
typedef std::function<void(const char* line)> snoop_callback_t;

static snoop_callback_t snoopCallback;


static void debug_snoop_putc(char c)
{
  uart_debug_putc(c);

  static char line[16];
  static unsigned lineLength = 0;

  if(c == '\r' || c == '\n') {
    line[lineLength] = '\0';
    snoopCallback(line);
    lineLength = 0;
  } else {
    if(lineLength < sizeof(line) - 1) {
      line[lineLength] = c;
      ++lineLength;
    }
  }
}

void setDebugSnoop(snoop_callback_t callback)
{
    snoopCallback = callback;
    
    ets_install_putc1(snoopCallback);
}

```


And in our debugging application we'd have:


```
void init()
{
	Serial.begin(SERIAL_BAUD_RATE);
	Serial.systemDebugOutput(true); // Allow debug print to serial

  setDebugSnoop([](const char* line) {
    if(memcmp(line, "E:M", 2) == 0) {
      print_stack();
    }
  });
  
  ...
  
}
```

**Comment by slaff** [comment]
> OK, so instead of changing `uart.cpp` we could ... Something like:

I like that!

**Comment by mikee47** [comment]
> > OK, so instead of changing `uart.cpp` we could ... Something like:
> 
> I like that!

How did I know you were going to say that. LOL.

**Comment by slaff** [comment]
> How did I know you were going to say that. LOL.

;-) Some might call it "magic" but I guess it is due to "consistency". A consistent behaviour leads to predictable behaviour same as with programming code. Thus it makes it easy to guess, use and advance.

**Comment by mikee47** [comment]
That's deep, man :-)

**Comment by mikee47** [comment]
OK, this is a more generic hook which, when installed, suppresses actual output so can do stuff like this:

```
	static char buffer[256];
	sysdbgInstallHook([](const char* line, unsigned length) {
		debug_w("[SYS] %s", line);
		if(memcmp(line, "E:M ", 4) == 0) {
			print_stack();
		}
	}, buffer, sizeof(buffer));
```

That includes the timestamp on system messages as well as trapping E:Mstuff. I've also blocked recursion in case further system messages get emitted during the callback.

**Comment by slaff** [comment]
@mikee47 Thanks for your efforts. Your comments are quite helpful but I would like to ask you to either modify an existing sample or create a new one demonstrating the usage of the new feature(s).

**Comment by mikee47** [comment]
> ... either modify an existing sample or create a new one demonstrating the usage of the new feature(s).

@slaff I'm thinking an addition to the LiveDebug sample, an additional command to force an "E:M" output, etc.

**Comment by slaff** [comment]
>> ... either modify an existing sample or create a new one demonstrating the usage of the new feature(s).

>  I'm thinking an addition to the LiveDebug sample, an additional command to force an "E:M" output, etc.

Sounds good to me.

**Comment by slaff** [comment]
@mikee47 Is the PR tested and ready to merge?

**Comment by mikee47** [comment]
@slaff It works, but taking another look at the code as I'm not completely happy with it yet. It feels clunky and overly complex to use. I'll update later today, but in the meantime if you get a chance to run it would welcome your feedback on how it fits in with debugging in general, and whether it's clear enough what it's supposed to be doing.

**Comment by mikee47** [comment]
I'm also not 100% happy with the name - `SystemDebug` would be better than `sysdbg` but both could be misleading; maybe `OsPrintfHook` or something?

**Comment by slaff** [comment]
> It feels clunky and overly complex to use. I'll update later today,

Perfect. Thanks in advance.

> maybe OsPrintfHook or something?

How about `OsMessageInterceptor`  or `SdkMessageInterceptor`?

**Comment by mikee47** [comment]
`OsMessageInterceptor` it is!

The rewrite is, I hope, an improvement. However, it does highlight an issue I've found using GDB when inspecting the stack. I'll illustrate with an example.

```
Remote debugging using /dev/ttyUSB0
gdbstub_init () at /media/r/sming-v4/Sming/gdb/gdbstub.cpp:908
908			gdbstub_do_break();
(gdb) c
Continuing.
LiveDebug sample
Explore some capabilities of the GDB debugger.

[OS] mode : softAP(de:4f:22:1c:26:66)
[OS] add if1
[OS] dhcp server start:(ip:192.168.4.1,mask:255.255.255.0,gw:192.168.4.1)
[OS] bcn 100
3083430 GdbAttach(1)
3138229 open log 0
(Attached) malloc0
6037893 gdb_read_console() returned 8
Attempting to allocate a zero-length array results in an OS debug message.
The message starts with 'E:M ...' and can often indicate a more serious memory allocation issue.
[OS] E:M 0
** OS Memory Error **

Program received signal SIGTRAP, Trace/breakpoint trap.
gdb_do_break () at /media/r/sming-v4/Sming/gdb/gdbstub.cpp:937
937		gdbstub_do_break();
```

All good so far. Now let's see if we can find which piece of code caused the problem.

```
(gdb) bt
#0  gdb_do_break () at /media/r/sming-v4/Sming/gdb/gdbstub.cpp:937
#1  0x402350f1 in onOsMessage (msg=...) at app/application.cpp:401
#2  0x4023fe5d in OsMessageInterceptor::putc (this=0x3ffee884 <osMessageInterceptor>, c=<optimized out>) at system/OsMessageInterceptor.cpp:23
#3  0x40237da8 in OsMessageInterceptor::__lambda0::operator() (__closure=0x0, c=<optimized out>) at system/OsMessageInterceptor.cpp:43
#4  OsMessageInterceptor::__lambda0::_FUN (c=<optimized out>) at system/OsMessageInterceptor.cpp:43
#5  0x40001db4 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb) 
```

It stops before we get to anything useful. The 'corrupt stack' warning doesn't help either, and I cannot find any way to get GDB to skip the dodgy stack frames. To get any further I have to skip ahead manually, thus:

```
(gdb) set $sp=$sp+0x100
(gdb) bt
#0  gdb_do_break () at /media/r/sming-v4/Sming/gdb/gdbstub.cpp:937
#1  0x402350f1 in onOsMessage (msg=...) at app/application.cpp:401
#2  0x40237eac in Print::println (this=0x3ffee884 <osMessageInterceptor>) at Wiring/Print.cpp:123
#3  0x402340ac in handleCommand_malloc0 () at app/application.cpp:415
(gdb) 
```

That's better.

```
(gdb) frame 3
#3  0x402340ac in handleCommand_malloc0 () at app/application.cpp:415
415		os_malloc(0);
(gdb) disass/m
Dump of assembler code for function handleCommand_malloc0():
410	{
   0x40234080 <+0>:	movi	a9, 192
   0x40234083 <+3>:	sub	a1, a1, a9
   0x40234086 <+6>:	s32i	a0, a1, 188

411		Serial.println(
412			_F("Attempting to allocate a zero-length array results in an OS debug message.\r\n"
   0x40234089 <+9>:	mov.n	a2, a1
   0x4023408b <+11>:	l32r	a3, 0x40234078
   0x4023408e <+14>:	movi	a4, 176
   0x40234091 <+17>:	call0	0x40236248 <memcpy_aligned(void*, void const*, unsigned int)>

413			   "The message starts with 'E:M ...' and can often indicate a more serious memory allocation issue."));
   0x40234094 <+20>:	l32r	a2, 0x40218cb4
   0x40234097 <+23>:	mov.n	a3, a1
   0x40234099 <+25>:	call0	0x40237ee4 <Print::println(char const*)>

414	
415		os_malloc(0);
   0x4023409c <+28>:	movi.n	a2, 0
   0x4023409e <+30>:	l32r	a3, 0x4023407c
   0x402340a1 <+33>:	movi	a4, 0x19f
   0x402340a4 <+36>:	movi.n	a5, 1
   0x402340a6 <+38>:	l32r	a0, 0x4020279c
   0x402340a9 <+41>:	callx0	a0

416	
417		return true;
418	}
=> 0x402340ac <+44>:	movi.n	a2, 1
   0x402340ae <+46>:	l32i	a0, a1, 188
   0x402340b1 <+49>:	movi	a9, 192
   0x402340b4 <+52>:	add.n	a1, a1, a9
   0x402340b6 <+54>:	ret.n

End of assembler dump.
(gdb) 
```

**Comment by mikee47** [comment]
Also, I note that the Arduino project uses a different java-based stack dumper (repo [here](https://github.com/me-no-dev/EspExceptionDecoder)), and uses GDB to get the address information (instead of addr2line).

**Comment by slaff** [comment]
> java-based stack dumper

Yes, the stack python stack dumper that we use is based on that code. I could check later on their progress and try to update the python code. 

> The 'corrupt stack' warning doesn't help either, and I cannot find any way to get GDB to skip the dodgy stack frames.

@igrr @earlephilhower  Do you have some ideas? (See the previous comment in the thread)

**Comment by igrr** [comment]
`0x40001db4` looks like it is ets_printf from ROM, so i assume that 1) GDB session doesn't have the ELF file for the ROM code and 2) ROM wasn't compiled with -fno-omit-frame-pointer.

**Comment by mikee47** [comment]
> `0x40001db4` looks like it is ets_printf from ROM, so i assume that 1) GDB session doesn't have the ELF file for the ROM code and 2) ROM wasn't compiled with -fno-omit-frame-pointer.

@igrr Thanks for the pointer - missing symbols it was

**Comment by mikee47** [comment]
@slaff With #1665 we now get:

```
(gdb) bt
#0  gdb_do_break () at r:/sming-v4/Sming/gdb/gdbstub.cpp:937
#1  0x40234f95 in onOsMessage (msg=...) at app/application.cpp:401
#2  0x4023de95 in OsMessageInterceptor::putc (this=0x3ffee884 <osMessageInterceptor>, c=<optimized out>)
    at system/OsMessageInterceptor.cpp:23
#3  0x40239728 in OsMessageInterceptor::static_putc (c=<optimized out>) at system/include/OsMessageInterceptor.h:93
#4  0x40001db4 in ets_write_char ()
#5  0x40001f46 in ets_vprintf ()
#6  0x40002514 in ets_printf ()
#7  0x40100650 in pvPortMalloc ()
#8  0x40233f64 in handleCommand_malloc0 () at app/application.cpp:415
#9  0x402348ac in handleCommand (cmd=...) at app/application.cpp:479
#10 0x40234b9d in onConsoleReadCompleted (info=...) at app/application.cpp:514
#11 0x4023682e in SystemClass::taskHandler (event=<optimized out>) at SmingCore/Platform/System.cpp:34
#12 0x40000f49 in ets_post ()
(gdb)
```

**Comment by mikee47** [comment]
I tried this again using the 'hang' command in `LiveDebug` sample:

```
(Attached) hang
590597360 gdb_read_console() returned 5
Entering infinite loop...


***** Software Watchdog Reset

Program received signal SIGPWR, Power fail/restart.
0x402352c3 in debug_crash_callback (rst_info=rst_info@entry=0x3ffffc10, stack=stack@entry=1073741248,
    stack_end=stack_end@entry=1073741744) at r:/sming-v4/Sming/appspecific/gdb/gdb_hooks.cpp:104
104             gdbstub_break_internal(DBGFLAG_RESTART);
(gdb) bt
#0  0x402352c3 in debug_crash_callback (rst_info=rst_info@entry=0x3ffffc10, stack=stack@entry=1073741248,
    stack_end=stack_end@entry=1073741744) at r:/sming-v4/Sming/appspecific/gdb/gdb_hooks.cpp:104
#1  0x40235bea in __wrap_system_restart_local () at system/crash_handler.c:66
#2  0x40103ed2 in pp_soft_wdt_feed_local ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb)
```

Looks like `pp_soft_wdt_feed_local` is defined in `libpp.a`. Any clues?

**Pull request merged**